src/components/button/template.njk

```diff
@@ -1,13 +1,37 @@
-{% if params.url %}
-<a class="govuk-c-button
-{%- if params.isStart %} govuk-c-button--start{% endif %}
-{%- if params.classes %} {{ params.classes }}{% endif %}" href="{{ params.url | safe }}" role="button">
-  {%- if params.text %}{{ params.text }}{% endif %}
-</a>
+{# Determine type of element to use, if not explicitly set -#}
+
+{% if params.element %}
+  {% set element = params.element %}
 {% else %}
-<input class="govuk-c-button
-{%- if params.isDisabled %} govuk-c-button--disabled{% endif %}
-{%- if params.classes %} {{ params.classes }}{% endif %}"
-{%- if params.text %} value="{{ params.text }}"{% endif %}
-{%- if params.isDisabled %} disabled="disabled" aria-disabled="true"{% endif %}>
+  {% if params.href %}
+    {% set element = 'a' %}
+  {% elseif params.html %}
+    {% set element = 'button' %}
+  {% else %}
+    {% set element = 'input' %}
+  {% endif %}
 {% endif %}
+
+{#- Define common attributes that we can use across all element types #}
+
+{%- set commonAttributes %} class="govuk-c-button{% if params.classes %} {{ params.classes }}{% endif %}{% if params.disabled %} govuk-c-button--disabled{% endif %}"{% for attribute, value in params.attributes %} {{attribute}}="{{value}}"{% endfor %}{% endset %}
```

Comment:
would it makes sense to add a check if param.href apply govuk-c-link--as-button so that in the link styles we can apply corrections? This is effectively a link variant:)

Reply:
We should be able to add corrections based on `a.govuk-c-button`?

Reply:
Sorry, just re-read this. I don't think we should see this a `link` variant, that just confuses things. It's a button that happens to be an `<a>`.

Reply:
yeah, just in terms of its BEM modifier. but can't think of any clever modifier names